lib/Backends/OpenCL/kernels_fwd_conv.cl

```diff
@@ -319,4 +335,221 @@ conv_forward_mem(__global void *mem, unsigned im_in_offset, unsigned wg_offset,
+              tiledIndex = tiledIndex / v_k_0;
+              // Compute d_temp.
+              // Scale d_temp by the stride and subtract the padding.
+              d_temp_0 = (imageIndex % v_imso_0) * v_s_0 - v_p_0;
+              imageIndex = imageIndex / v_imso_0;
+              // Recombine final index, compute in-range.
+              bool skip_range_check = false;
+              // Used only if padding is not 0.
+              bool in_range = !skip_range_check;
+              int_tp d_iter_im;
+              // Here, d_temp_ represents the column shift,
+              // while d_iter_ is the kernel shift.
+              d_iter_im = d_temp_0 + d_iter_0;
+              tiledIndex = tiledIndex * v_imsi_0 + d_iter_im;
+              if (!skip_range_check) {
+                in_range &= d_iter_im >= 0 && d_iter_im < v_imsi_0;
+              }
+              // Here, d_temp_ represents the column shift,
+              // while d_iter_ is the kernel shift.
+              d_iter_im = d_temp_1 + d_iter_1;
+              tiledIndex = tiledIndex * v_imsi_1 + d_iter_im;
+              if (!skip_range_check) {
+                in_range &= d_iter_im >= 0 && d_iter_im < v_imsi_1;
+              }
+              if (skip_range_check || in_range) {
+                // tiledIndex now holds the memory offset for the input image.
+                Bsub[row][col] = Bptr[tiledIndex];
+              } else {
+                Bsub[row][col] = ZERO;
+              }
+            }
+          }
+        }
+        // Synchronize to make sure the tile is loaded.
+        barrier(CLK_LOCAL_MEM_FENCE);
+        // Temporary registers for A and B.
+        Dtype4 Areg;
+        Dtype4 Breg[WPTN / VWN];
+// Loop over the values of a single tile.
+#pragma unroll 1
+        for (int_tp kt = 0; kt < TSK; kt += TSK_UNROLL) {
+#pragma unroll 1
+          for (int_tp ku = 0; ku < TSK_UNROLL; ++ku) {
+            int_tp k = kt + ku;
+// Cache the values of Bsub in registers.
+#pragma unroll
+            for (int_tp wn = 0; wn < WPTN / VWN; ++wn) {
+              int_tp col = tidn + wn * VWN * RTSN;
+              VEC_4_0(Breg[wn]) = Bsub[k][col + 0 * RTSN];
+              VEC_4_1(Breg[wn]) = Bsub[k][col + 1 * RTSN];
+              VEC_4_2(Breg[wn]) = Bsub[k][col + 2 * RTSN];
+              VEC_4_3(Breg[wn]) = Bsub[k][col + 3 * RTSN];
+            }
+// Perform the computation.
+#pragma unroll
+            for (int_tp wm = 0; wm < WPTM / VWM; ++wm) {
+              int_tp row = tidm + wm * VWM * RTSM;
+              VEC_4_0(Areg) = Asub[row + 0 * RTSM][k];
+              VEC_4_1(Areg) = Asub[row + 1 * RTSM][k];
+              VEC_4_2(Areg) = Asub[row + 2 * RTSM][k];
+              VEC_4_3(Areg) = Asub[row + 3 * RTSM][k];
+#pragma unroll
+
+
+              for (int_tp wn = 0; wn < WPTN / VWN; ++wn) {
+                VEC_4_0(Creg[wm * VWM + 0][wn]) +=
+                    A_OFFSET(VEC_4_0(Areg)) * B_OFFSET(VEC_4_0(Breg[wn]));
```

Comment:
It seems like there are a lot of repeating A_OFFSET or B_OFFSET computations with the same value of the argument. It could be a good idea to compute them just once. It is probably easier to actually perform these computations when initializing the values of `Areg` and `Breg` above. Or even when initializing the values of `Asub` and `Bsub`.